Update documentation tooling

utzig opened the pull request:
Update documentation build tools to latest versions: Breathe==4.20.0 (released last week) and Sphinx>=3.2.0.

Since the long standing bug that disallowed using the C domain was fixed recently, switch to using it by updating all :cpp: references to equivalent :c: references. This way only the C parser in Sphinx is used and there is one can just use :c: with the type of the identifier being referenced (struct, type, func, enum, member, etc).

Commit:
doc: use non-deprecated CSS/JS import methods

Update methods used to include JS/CSS files, avoiding deprecated
versions the don't work on Sphinx>=3.x. The new methods work on both
Sphinx==3.x and Sphinx>=2.4.x.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: switch from CPP to C domain for .c files

Older Breathe versions didn't work correctly with the C domain, so there
was a mismatch when refering to functions and #defines, using :cpp: or
:c: depending on the origin of the refered symbol. With Breathe >=4.19
the C domain works correctly so all symbols can be refered to with the
:c: role.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: move tabs to spaces in Sphinx config

Sphinx's config file is a python script, so follow the standards and use
spaces for identation.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: add __subsystem attribute and c attributes

Add __subsystem to list of attributes for the parser, and make the c
parser accept the same attributes than the cpp parser, after the change
to use the c parser for ".c" files.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: do not use :c:macro: for NULL

`NULL` is not defined in the Zephyr sources, but rather a C definition
which is not resolved when building the documentation and generates a
warning. Fall back to using just emphasis.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: fix invalid usage of function to :c:macro:

Some macros were using :c:func: and :cpp:func: which forbids Breathe
from resolving those symbols. Update to :c:macro: for proper resolution.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: update enum usage to use :c:enum:

Update enum reference usage to use C domain.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: update union reference to to :c:union

Update union reference from generic :c:type: to :c:union:.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: update :cpp:var: to :c:member:

Update references to struct members to use :c:member:

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: update versions of build tools

Update documentation build tooling:

- Sphinx>=3.2.0
- Breathe>=4.20.0
- docutil>=0.16.0

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: update struct references to use :c:struct:

Sphinx>=3.0 includes the `struct` role in the C domain, which provides a
specific way to link to structs, while the old :c:type: role should be
primary used to typedefs. Update existing references, using :c:type:,
:cpp:type: or emphasized symbols that point to structs to use the new
role.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: add :c:macro to some symbols

Fix some C macros that are currently just emphasized in the
documentation with proper symbol resulition using :c:macro:

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: update enumerator usage to :c:enumerator:

Following the change to use the C domain for parsing, update all
existing :cpp:enumerator: references to :c:enumerator:.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: update function references to use :c:func:

Following the change to use the C domain for parsing, update all
existing :cpp:func: references to :c:func:. Remove the parentheses as
well, if used, because they are not needed, this is already known to be
a function, and how it is displayed in the documentation later is a
semantic decision done by the output builder.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

nashif commented on .github/workflows/doc-build.yml:
```diff
@@ -30,8 +30,8 @@ jobs:
         pip3 install setuptools
-        pip3 install 'breathe>=4.9.1,<4.15.0' 'docutils>=0.14' \
-                     'sphinx>=1.7.5,<3.0' sphinx_rtd_theme sphinx-tabs \
+        pip3 install 'breathe>=4.20.0' 'docutils>=0.16' \
```
maybe just use `pip3 install -r scripts/requirements-doc.txt`, this way we do not have to maintain the versions twice and get out of sync.

utzig replied:
done!

nashif commented on .github/workflows/doc-publish.yml:
```diff
@@ -61,8 +61,8 @@ jobs:
         pip3 install setuptools
-        pip3 install 'breathe>=4.9.1,<4.15.0' 'docutils>=0.14' \
-                     'sphinx>=1.7.5,<3.0' sphinx_rtd_theme sphinx-tabs \
+        pip3 install 'breathe>=4.20.0' 'docutils>=0.16' \
```
ditto

utzig replied:
ditto!

Commit:
doc: update known issues for Sphinx>=3.2.0

Remove old known issues files and update with issues that still need to
be fixed in Sphinx>=3.2.0 and or Doxygen, including anonymous
union/struct parsing and duplicated symbols (eg struct and functions
having the same name).

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

Commit:
doc: fix :c:macro: usage

Remove parameters in :c:macro: usage because it's invalid syntax.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

chrta commented on scripts/requirements-doc.txt:
```diff
@@ -1,8 +1,8 @@
-sphinx>=1.7.5,<3.0
+breathe>=4.20.0
+docutils>=0.16
+sphinx>=3.2.0
```
Only requiring >=3.2.0 for sphinx seems dangerous to me. In the past sphinx changed a lot in their api that broke compatibility with extensions and themes when updating major versions.

utzig replied:
What is your suggestion?

chrta replied:
Maybe `sphinx>=3.2.0,<4.0` would be good for now.

utzig replied:
Ok, the latest release is `3.2.1` and there are no development releases with updated major versions, so it's really hard to know right now if they'll be compatible or not, but I will update as suggested.

de-nordic replied:
I do not think this is an issue. If `sphinx` follows semantic versioning (https://semver.org/) then change in major version is exactly doing what might have been expected.

de-nordic replied:
I have misunderstood the problem. Sorry.

utzig replied:
I changed it to what @chrta suggested; my personal opinion would be to not set upper limits right now, but they won't really affect us for a few months at least.

utzig commented:
I removed the DNM now because I am fairly happy with the results when compared to the official docs. I've not seen a single link broken that was not already broken before, and this PR is a huge improvement in this regard, eg, check this section: https://docs.zephyrproject.org/latest/reference/usermode/memory_domain.html#thread-resource-pools. Every API link is broken there and almost all of them works on this PR. This also applies to every pages touched by this PR that I reviewed manually.

chrta reviewed (changes requested):
I am currently building the doc to approve this later. I noticed that the version information in zephyr/doc/README.rst should also be updated, see https://docs.zephyrproject.org/latest/README.html#installing-the-documentation-processors

Edit: The generated doc looks good :+1:

utzig commented:
> I am currently building the doc to approve this later. I noticed that the version information in zephyr/doc/README.rst should also be updated, see https://docs.zephyrproject.org/latest/README.html#installing-the-documentation-processors

Done; I squashed the update into the first commit that updates the tooling.
